python/GafferUI/MessageWidget.py

```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+			self._qtWidget().selectAll()
+			return True
+
+		elif event.key == "F" and event.modifiers == event.Modifiers.Control and self.__searchWidget is not None :
+
+			self.__searchWidget.grabFocus()
+			return True
+
+		elif event.key in self.__eventLevelShortcuts :
+
+			if event.modifiers == event.Modifiers.None_ :
+				self.nextMessage( self.__eventLevelShortcuts[ event.key ] )
+				return True
+			elif event.modifiers == event.Modifiers.Shift :
+				self.previousMessage( self.__eventLevelShortcuts[ event.key ] )
+				return True
+
+		elif event.key == "S" and event.modifiers == event.Modifiers.None_ :
+
+			self.setFocusSearchResults( not self.getFocusSearchResults() )
+			return True
+
+		elif event.key == "P" and event.modifiers == event.Modifiers.None_ :
+
+			self.previousSearchResult()
+			return True
+
+		elif event.key == "N" and event.modifiers == event.Modifiers.None_ :
+
+			self.nextSearchResult()
+			return True
+
+		elif event.key in ( "End", "B" ) and event.modifiers == event.Modifiers.None_ :
+
+			self.__userSetScrollPosition( False )
+			self.__scrollToBottom()
+			return True
+
+		return False
+
+	# Copy/Paste
+
+	def __copySelectedRows( self ) :
+
+		# TODO only slected, can we get something for free from QT?
+
+		messageIndices = self.__selectedMessageIndices()
+		text = self.__plainTextForMessages( messageIndices )
+		QtWidgets.QApplication.clipboard().setText( text )
+
+	def __plainTextForMessages( self, messageIndices ) :
+
+		messages = self.getMessages()
+		indices = messageIndices or range( len(messages) )
+
+		text = ""
+		for i in indices :
+			m = messages[ i ]
+			text += "%s [%s] %s\n" % ( str(m.level).ljust(7).upper(), m.context, m.message )
+
+		return text
+
+# Expands messages into a two-row presentation, with level + context on one
+# row, and the body of the message on the next.
+class _MessageTableExpandedViewProxy( QtCore.QAbstractProxyModel ) :
+
+	def columnCount( self, parent ) :
+
+		return 1
+
+	def rowCount( self, parent ) :
+
+		if parent.isValid() :
+			return 0
+
+		return 2 * self.sourceModel().rowCount() if self.sourceModel() else 0
+
+	def mapFromSource( self, sourceIndex )  :
+
+		if not sourceIndex.isValid() or sourceIndex.row() < 0 :
+			return QtCore.QModelIndex()
+
+		row = sourceIndex.row() * 2
+		if sourceIndex.column() == int( _MessageTableModel.Column.Message ) :
+			row += 1
+
+		return self.index( row, 0 )
+
+	def mapToSource( self, proxyIndex ) :
+
+		if not proxyIndex.isValid() or proxyIndex.row() < 0 :
+			return QtCore.QModelIndex()
+
+		if proxyIndex.row() % 2 == 0 :
+			column = _MessageTableModel.Column.Level
+		else :
+			column = _MessageTableModel.Column.Message
+
+		return self.sourceModel().index( proxyIndex.row() // 2, int(column) )
+
+	def mapSelectionFromSource( self, sourceSelection ) :
+
+		selection = QtCore.QItemSelection()
+		for i in sourceSelection.indexes() :
+			index = self.mapFromSource( i )
+			selection.select( index, index )
+		return selection
```

Comment:
I'm probably missing something because I'm reading about proxy models while reviewing this, but how is this different from the default implementation provided by the base class? And if we need to override `mapSelectionFromSource`, why don't we need to override `mapSelectionToSource()`?

Reply:
Good spot, thanks. These were a hang over from 'when nothing was working'. https://github.com/GafferHQ/gaffer/pull/3790/commits/aa07f1f32c4299cdc55ec98691f3a7d27cc06be7